src/basic/user-util.c

```diff
@@ -555,18 +557,11 @@ int getgroups_alloc(gid_t** gids) {
                 ngroups = getgroups(0, NULL);
                 if (ngroups < 0)
                         return -errno;
-                if (ngroups == 0)
-                        return false;
-
-                free(allocated);
-
-                p = allocated = new(gid_t, ngroups);
-                if (!allocated)
-                        return -ENOMEM;
+                if (ngroups == 0) {
+                        *ret = NULL;
+                        return 0;
+                }
```

Comment:
This is inconsistent with the behavior from right above, which returns a non-NULL pointer to an empty buffer on the first iteration.  (New lines 544-547.)

It is also inconsistent with the fact that this project uses the `?:1` trick in malloc(3)/realloc(3) to avoid annoying NULLs messing around.

```c
alx@devuan:~/src/systemd/systemd/agetgroups$ find -type f | grep '\.[ch]$' | xargs grepc new
./src/basic/alloc-util.h:#define new(t, n) ((t*) malloc_multiply(n, sizeof(t)))
alx@devuan:~/src/systemd/systemd/agetgroups$ sed -n 115,120p src/basic/alloc-util.h
_malloc_ _alloc_(1, 2) static inline void *malloc_multiply(size_t need, size_t size) {
        if (size_multiply_overflow(size, need))
                return NULL;

        return malloc(size * need ?: 1);
}
```

Reply:
Plus, this is fixing a double-free bug, and doesn't acknowledge it.

Reply:
> This is inconsistent with the behavior from right above, which returns a non-NULL pointer to an empty buffer on the first iteration. (New lines 544-547.)

Yes, this is inconsistent but in some very impossible way, because if the list is empty then the first call would already have succeeded and thus if the list is empty we never actually should hit this 2nd code path here.

But I'll grant you that it would be better to be consistent. Which means to return a NULL list too on the first call if 0 is returned.

generally, in the systemd codebase we try to treat NULL and empty arrays mostly the same, to simplify things and optimize things